File: internal/message/subscribe_update.go
package message

import (
	"io"
	"log/slog"
	"time"

	"github.com/quic-go/quic-go/quicvarint"
)

type SubscribeUpdateMessage struct {
	SubscribeID SubscribeID

	SubscriberPriority SubscriberPriority
	GroupOrder         GroupOrder
	GroupExpires       time.Duration
	MinGroupSequence   GroupSequence
	MaxGroupSequence   GroupSequence

	Parameters Parameters
}

func (su SubscribeUpdateMessage) Encode(w io.Writer) error {
	slog.Debug("encoding a SUBSCRIBE_UPDATE message")
	/*
	 * Serialize the message in the following format
	 *
	 * SUBSCRIBE_UPDATE Message {
	 *   Subscribe ID (varint),
	 *   Subscriber Priority (byte),
	 *   Min Group Number (varint),
	 *   Max Group Number (varint),
	 *   Subscribe Update Parameters (Parameters),
	 * }
	 */

	/*
	 * Serialize the payload
	 */
	p := make([]byte, 0, 1<<6)

	// Append the Subscriber ID
	p = quicvarint.Append(p, uint64(su.SubscribeID))

	// Append the Subscriber Priority
	p = quicvarint.Append(p, uint64(su.SubscriberPriority))

	// Append the Min Group Number
	p = quicvarint.Append(p, uint64(su.MinGroupSequence))

	// Append the Max Group Number
	p = quicvarint.Append(p, uint64(su.MaxGroupSequence))

	// Append the Subscribe Update Parameters
	p = appendParameters(p, su.Parameters)

	// Get a serialized message
	b := make([]byte, 0, len(p)+8)

	// Append the length of the payload
	b = quicvarint.Append(b, uint64(len(p)))

	// Append the payload
	b = append(b, p...)

	// Write
	_, err := w.Write(b)

	return err
}

func (sum *SubscribeUpdateMessage) Decode(r io.Reader) error {
	slog.Debug("decoding a SUBSCRIBE_UPDATE message")
	// Get a messaga reader
	mr, err := newReader(r)
	if err != nil {
		return err
	}
	// Get a Subscribe ID
	num, err := quicvarint.Read(mr)
	if err != nil {
		return err
	}
	sum.SubscribeID = SubscribeID(num)

	// Get a Min Group Number
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	sum.MinGroupSequence = GroupSequence(num)

	// Get a Max Group Number
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	sum.MaxGroupSequence = GroupSequence(num)

	// Get a Subscriber Priority
	priorityBuf := make([]byte, 1)
	_, err = r.Read(priorityBuf)
	if err != nil {
		return err
	}
	sum.SubscriberPriority = SubscriberPriority(priorityBuf[0])

	// Get Subscribe Update Parameters
	sum.Parameters, err = readParameters(mr)
	if err != nil {
		return err
	}

	return nil
}

File: moq-web/src/message/subscribe_update.ts
import type { Writer, Reader } from "../io.ts";
import { varintLen } from "../io/len.ts";
import type { GroupSequence } from "../protocol.ts";

export interface SubscribeUpdateMessageInit {
    trackPriority?: number;
    minGroupSequence?: GroupSequence;
    maxGroupSequence?: GroupSequence;
}

export class SubscribeUpdateMessage {
    trackPriority: number;
    minGroupSequence: GroupSequence;
    maxGroupSequence: GroupSequence;

    constructor(init: SubscribeUpdateMessageInit) {
        this.trackPriority = init.trackPriority ?? 0;
        this.minGroupSequence = init.minGroupSequence ?? 0n;
        this.maxGroupSequence = init.maxGroupSequence ?? 0n;
    }

    get messageLength(): number {
        return (
            varintLen(this.trackPriority)
             + varintLen(this.minGroupSequence)
             + varintLen(this.maxGroupSequence)
        );
    }

    async encode(writer: Writer): Promise<Error | undefined> {
        writer.writeVarint(this.messageLength);
        writer.writeVarint(this.trackPriority);
        writer.writeBigVarint(this.minGroupSequence);
        writer.writeBigVarint(this.maxGroupSequence);
        return await writer.flush();
    }

    async decode(reader: Reader): Promise<Error | undefined> {
        let [len, err] = await reader.readVarint();
        if (err) {
            return err;
        }
        [this.trackPriority, err] = await reader.readVarint();
        if (err) {
            return err;
        }
        [this.minGroupSequence, err] = await reader.readBigVarint();
        if (err) {
            return err;
        }
        [this.maxGroupSequence, err] = await reader.readBigVarint();
        if (err) {
            return err;
        }

        if (len !== this.messageLength) {
            throw new Error(`message length mismatch: expected ${len}, got ${this.messageLength}`);
        }

        return undefined;
    }
}